wordpress-mobile/WordPress-Android
# Issue/13066 ppn categories

Part of #13066 

This PR adds support for
1. Selecting categories for a post
2. Adding a new category

This PR is labeled WIP/do not merge. I am looking for a review before moving forward with the open issues (see list below).

The meat of the PR can be found in:
`PrepublishingAddCategoryFragment` and `PrepublishingAddCategoryViewModel`
- This pair is responsible for the adding of a new category 

`PrepublishingCategoriesFragment and PrepublishingCategoriesViewModel`
- This pair is responsible for selecting/unselecting categories and for kicking off the Add Category action.

`PrepublishingBottomSheetFragment` (preexisting) is the traffic cop and handles adding/removing fragment transactions to the bottom sheet container. This was updated to handle the AddCategory and Categories fragments. Each of the individual actions in the bottom sheet are standalone fragments and the behavior is identical for all.

I was on the fence with regards to suspendable functions.
- They are used for update/add/fetch server calls in the `UpdatePostCategoriesUseCase`, `AddCategoryUseCase`, and `GetCategoriesUseCase`. I followed existing architecture for calls to TaxonomyStore and EditPostRepository. I am open to suggestions.

It’s a rather larger PR, but I broke it up by commits.
f8a263b => layout files, animation files, strings, `AppComponent`/`ViewModelModule`, plugging into existing architecture
740ce31 => Changes to `PrepublishingBottomSheetFragmentPrepublishingHomeViewModel`, `PrepublishingViewModel`
7c5012d => Add category related
3d28bb4 => Select category related

Home|Select|Add
---|---|---
<img width="480" src="https://user-images.githubusercontent.com/506707/96194465-6995f880-0f18-11eb-8959-d21eea29fd6f.png">|<img width="480" src="https://user-images.githubusercontent.com/506707/96194480-73b7f700-0f18-11eb-9202-25a09d2ba0a2.png">|<img width="480" src="https://user-images.githubusercontent.com/506707/96194499-7c103200-0f18-11eb-8825-2d9cca288bdf.png">

**This PR has open issues** and they are almost all related to BottomSheet behavior (scrolling the biggest one)
1. Showing a snackbar in a bottom sheet (using toast for now)
~~2. Recyclerview: scrolling within a bottom sheet (this also affects PTR)~~  ****Update**: Removed the PTR in commit cd27ec0 and scrolling works just fine. 
3. There is an existing [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/12166) for the janky transitions between the bottom sheet fragments. I started on some new transition files and things were looking positive, but ultimately decided to push this issue off until all the bottom sheet scrolling issues are resolved.
4. There is a separate [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/13149) for writing unit test 
I am open to any suggestions for solving the scrolling/recyclerview issues in a bottom sheet. 😉 

To test:
Prep
- Launch the app
- Select Blog posts
- Select Edit on a post
- From the edit post view, tap the UPDATE action to launch the bottom sheet

Categories are shown on the Categories row
- If the post has categories assigned, you will see them listed on the home view (figure 1 above). If not, you will see Not Set

Select Categories is shown when user taps on Categories
- Tap the categories row to launch the Select categories view (fig 2)

Add category is shown and user can save a new category
- Tap the Add New Category button on the bottom of the select categories view (fig 3)
- Enter a new category and tap the <- Back Arrow

Newly Added category is shown in the categories list
- Take note of the toast which will tell you if the action succeeded or failed. 
- Back on the select category view, the category you have added is visible and checked

Categories can be checked/unchecked 
- Select or unselect some categories

Categories are saved
- Tap the <- Back arrow to save your options
- The categories row on the home screen should reflect the selections you made

PR submission checklist:
